Upgrade Electron to v1.6.x

thomasjo opened the pull request:
### Reported Issues

* [x] Object undefined errors from Electron https://github.com/atom/atom/pull/12696#issuecomment-261263408 and https://github.com/atom/atom/pull/12696#issuecomment-261527755
* [x] Default CSP changed, breaking file-icons and specs
* [x] `os.tmpDir()` deprecation warnings https://github.com/atom/atom/pull/12696#issuecomment-288871751
* [x] Drag image in `tree-view` incorrect (https://github.com/atom/tree-view/pull/1054)
* [x] Drop marker on `tabs` and `tree-view` missing (https://github.com/atom/tree-view/pull/1055, https://github.com/atom/tabs/pull/426)
* [x] Changes from https://github.com/atom/atom/pull/13795 cause exceptions in minimap (https://github.com/atom-minimap/minimap/pull/582)
* [x] Dragging pane items between docks mistakenly shows drop indicators due to new security constraints for the `dragover` event `dataTransfer` object (https://github.com/atom/tabs/pull/437)
* [x] Developer tools stop updating after reloading a window in dev mode
* [x] Issues folding soft-wrapped lines https://github.com/atom/atom/pull/12696#issuecomment-301161445
* [x] Dev tools reload exception (https://github.com/atom/atom/pull/12696#issuecomment-301207979, filed a bug for this in https://github.com/electron/electron/issues/9493)
* [x] Dev tools reload crash (end of https://github.com/atom/atom/pull/12696#issuecomment-301207979, filed a bug for this in https://github.com/electron/electron/issues/9493)
* [ ] White flash when hiding toolbar https://github.com/atom/atom/pull/12696#issuecomment-301182375
* [ ] Warnings from tree-view https://github.com/atom/atom/pull/12696#issuecomment-301208739
* [ ] Consistently crashing when continuing from a devtools breakpoint (filed a bug for this in https://github.com/electron/electron/issues/9078)

Reported issues resolved by #13880:

* [x] Scrollbars misbehaving on first file that is opened/restored https://github.com/atom/atom/pull/12696#issuecomment-251299649
* [x] Shift-scrolling scrolls vertically instead
* [x] Failing `spec/text-editor-component-spec.js` tests

---

During my initial testing on macOS 10.12, this seems to be working. Haven't noticed any regressions.
Would be great if more folks could give this a spin 🙇

/cc @atom/maintainers 

---

Fixes #11692 
Resolves #12690
Fixes #8006
Fixes #12832 
Fixes #13767 
Fixes #4084 (Will be fixed by the editor rendering rewrite)
Fixes #13265 (Will be fixed by the editor rendering rewrite)

thomasjo commented:
Since Electron v1.4.1 ships with https://github.com/electron/electron/pull/7209, I hope this PR will resolve #11692.

50Wliu commented:
Observations from around 10 minutes of light usage:
# NOTHING IS BLURRY ANYMORE!
I have beautiful subpixel antialiasing on everything, even the actual code for the first time ever\* :boom:.  Fixes #12652, maybe #7261, #3869, #2833, #2833.

Scrollbars seem to disappear randomly.  Haven't narrowed this down but I did notice one of my files didn't have a scrollbar.

\* Just noticed that the tabs and mini-editors don't have subpixel AA, but that's super minor compared to before.

50Wliu commented:
Ok, something else related to subpixel AA:
If I scroll all the way to the end of the file, I lose subpixel AA in the editor until I switch tabs.

Ben3eeE commented:
> Scrollbars seem to disappear randomly. Haven't narrowed this down but I did notice one of my files didn't have a scrollbar.

Have this happening with the vertical scrollbar on the TextEditor active when restoring session. In the case of split panes (I tried left/right) both TextEditors are missing vertical scrollbars.

1. Open Atom `1.12-dev-fcddf83` open some file that has content enough to have a vertical scrollbar.
2. Close Atom.
3. Open Atom `1.12-dev-fcddf83` and keep an eye on the vertical scrollbar. It appears quickly when opening the previous session and then disappears.

If you resize the pane so that the horizontal scrollbar appears or disappears then you can get the vertical scrollbar to appear, but in all other cases this file seems to be missing the vertical scrollbar. 

Tested with both Atom Light and One Dark themes. This does not reproduce in 1.11-beta5. Don't know about other dev versions.

50Wliu commented:
The disappearing scrollbar issue appears to happen on the first file that is opened/restored.  From the dev tools, the scrollbar (and scrollbar corner) is being given a width of 10px instead of the usual 15px.  If I change it to 15, then the scrollbar appears.  Anything > 10 works.  /cc @simurai
I've also been seeing a related issue where the scrollbar flickers like crazy when moving the mouse between the tab bar and the editor.  I'll try to find a stable repro case for that, but it also happens on the first file.

Now, some more subpixel antialiasing updates.  As I said before, it works for the most part.  Here's when it fails:
1. In a mini-editor
2. (Almost) anywhere with rendered text (eg Settings View, Tabs, Notifications, About).  The tree view and status bar are fine though?
3. Upon reaching the end of the file
4. Upon scrolling with a touchpad or scrollbar.  Mousewheel scrolling is fine.
5. Upon editing the last three visible lines, causing the editor to scroll and triggering case 4).

damieng commented:
Chrome tends to switch from sub-pixel to regular AA when it expects to need to redraw it several times such as in animation. This is because AA can be cached and re-used whereas sub-pixel depends on the horizontal offset.

simurai commented:
The scrollbar issue is super weird. Changing the `width` to anything other than the original (`10px`) works.. but when changing `bottom` it alternates. See the sub-pixel (`0.1px`) steps.

![scrollbars](https://cloud.githubusercontent.com/assets/378023/19063574/0c550f06-8a3f-11e6-9f20-68765e6750b6.gif)

simurai commented:
Another thing I noticed. The native tabs in macOS Sierra seem to work in this Electron version. Problem with the custom title bar is that the native tabs are just on top and don't push the rest down. Also, hard to read with dark themes.

![screen shot 2016-10-06 at 4 06 06 pm](https://cloud.githubusercontent.com/assets/378023/19143545/9821528c-8bdf-11e6-9d0d-c7c757df9f50.png)

Native title bar works fine.

### Steps to reproduce:

1. macOS Prefereces > Dock > "Prefer tabs when opening documents" to "Always":
  ![screen shot 2016-10-06 at 4 02 36 pm](https://cloud.githubusercontent.com/assets/378023/19143384/7dee1018-8bde-11e6-90df-023656923298.png)
2. Open another window with `cmd-o`.

### Possible solution

a. Either disable native tabs when custom title bar is enabled.
b. See if it's possible to detect when native tabs get added and then push the content down.

Maybe not _that_ urgent, since you have to enable two things to run into this. Could also be a separate issue.

thomasjo commented:
@simurai Electron v1.4.2 disabled the native tabs support, so this shouldn't be an issue anymore 🎱

Ben3eeE commented:
🙈 Drag image 🙈 

![dragimage](https://cloud.githubusercontent.com/assets/1058982/19199011/42edd086-8cc2-11e6-9a4d-43688deff500.gif)

v3ss0n commented:
electron 1.x is coming in atom 1.12?

50Wliu commented:
Correct.

Palabola commented:
Blurry Text still exist under Windows 7 even with the newest Electron!

Ben3eeE commented:
Fixes https://github.com/atom/atom/issues/13104.

Ben3eeE commented:
Seems the drop marker is missing again:
![drop marker](https://cloud.githubusercontent.com/assets/1058982/20317956/9dfa1cf4-ab68-11e6-8974-05763f4b4201.gif)
![drop marker 2](https://cloud.githubusercontent.com/assets/1058982/20317957/9dfd90fa-ab68-11e6-82f5-b4db4be4b589.gif)

nathansobo commented:
It would be helpful if we could assemble a list of issues that need to be resolved before we can merge this in the body of the issue.

Ben3eeE commented:
Updated the body with the issues I could find. @50Wliu if you could also check that would be great.

@50Wliu I can reproduce the scrollbar flickering issue when hovering over the tab bar. There have been a few reports of various things flickering when moving the mouse over the tab bar since 1.11. Maybe we do something strange?

damieng commented:
@Palabola can you attach a screenshot showing what you mean by blurry text?

50Wliu commented:
I've removed the dev tools freezing since I can't reproduce it anymore.

thomasjo commented:
I've narrowed down the problem with the missing drop indicators to being some problem related to `DataTransfer`.

The values set during the `dragstart` event via `event.dataTransfer.setData(format, value)` are empty strings during in the `dragover` event via `event.dataTransfer.getData(format)`.

Haven't yet figured out what has changed in Chromium and how to fix it. Quick, cursory search did not reveal any obvious Chromium bugs.

/cc @Ben3eeE @atom/maintainers

Ben3eeE commented:
Got this today when on this branch in dev mode.

![image](https://cloud.githubusercontent.com/assets/1058982/20393389/1262d418-acdc-11e6-9e29-b4ce8bf213c7.png)

Ben3eeE commented:
Another object of undefined error today when reloading the Atom window. I was fooling around with my `init.coffee` writing keystroke api code. Previous was when tabbing out of Atom. This time with Electron 1.4.7.

![image](https://cloud.githubusercontent.com/assets/1058982/20431051/2df92652-ad98-11e6-937a-97a3fdf1255a.png)

nathansobo commented:
@Ben3eeE I've added the object undefined errors to the list in the issue body.

kevinsawicki commented:
@damieng just a heads up, Electron 1.4.14 contains the `normalizeAccessKeys` option from https://github.com/electron/electron/pull/8308

Ben3eeE commented:
Two errors when testing latest artifact here. I don't see these in 1.15.0-beta3.

```
Failed to activate the status-bar package
contentWidth is not defined
Hide Stack Trace
ReferenceError: contentWidth is not defined
    at AtomClockView.adjustElementSize (atom-clock-view.js:97:5)
    at AtomClockView.initialize (atom-clock-view.js:21:10)
    at AtomClockView.start (atom-clock-view.js:14:10)
    at Object.consumeStatusBar (atom-clock.js:45:24)
    at Provider.module.exports.Provider.provide (C:\Users\lineri\AppData\Local\atom\app-1.16.0-dev-49ec99d\resources\app.asar\node_modules\service-hub\lib\provider.js:29:52)
    at ServiceHub.module.exports.ServiceHub.provide (C:\Users\lineri\AppData\Local\atom\app-1.16.0-dev-49ec99d\resources\app.asar\node_modules\service-hub\lib\service-hub.js:30:20)
    at Package.module.exports.Package.activateServices (C:\Users\lineri\AppData\Local\atom\app-1.16.0-dev-49ec99d\resources\app.asar\src\package.js:396:71)
    at Package.module.exports.Package.activateNow (C:\Users\lineri\AppData\Local\atom\app-1.16.0-dev-49ec99d\resources\app.asar\src\package.js:221:16)
    at C:\Users\lineri\AppData\Local\atom\app-1.16.0-dev-49ec99d\resources\app.asar\src\package.js:190:32
    at Package.module.exports.Package.measure (C:\Users\lineri\AppData\Local\atom\app-1.16.0-dev-49ec99d\resources\app.asar\src\package.js:96:15)
    at C:\Users\lineri\AppData\Local\atom\app-1.16.0-dev-49ec99d\resources\app.asar\src\package.js:183:26
    at Package.module.exports.Package.activate (C:\Users\lineri\AppData\Local\atom\app-1.16.0-dev-49ec99d\resources\app.asar\src\package.js:180:34)
    at PackageManager.module.exports.PackageManager.activatePackage (C:\Users\lineri\AppData\Local\atom\app-1.16.0-dev-49ec99d\resources\app.asar\src\package-manager.js:550:34)
    at C:\Users\lineri\AppData\Local\atom\app-1.16.0-dev-49ec99d\resources\app.asar\src\package-manager.js:531:29
    at Config.module.exports.Config.transactAsync (C:\Users\lineri\AppData\Local\atom\app-1.16.0-dev-49ec99d\resources\app.asar\src\config.js:337:18)
    at PackageManager.module.exports.PackageManager.activatePackages (C:\Users\lineri\AppData\Local\atom\app-1.16.0-dev-49ec99d\resources\app.asar\src\package-manager.js:526:19)
    at PackageManager.module.exports.PackageManager.activate (C:\Users\lineri\AppData\Local\atom\app-1.16.0-dev-49ec99d\resources\app.asar\src\package-manager.js:508:46)
    at C:\Users\lineri\AppData\Local\atom\app-1.16.0-dev-49ec99d\resources\app.asar\src\atom-environment.js:812:28
```

```
Uncaught TypeError: atom.views.pollDocument is not a function  C:\Users\lineri\.atom\packages\minimap\lib\minimap-element.js:345 
```

thomasjo commented:
@Ben3eeE Not seeing those errors on macOS 10.12.3, running Atom 1.16.0-dev-49ec99d3b.

Seems like the error is originating from minimap; I guess it's consuming the status-bar service, and the exception is causing a "catastrophic" error chain. Disable minimap to see if you can repro?

Ben3eeE commented:
I'm guessing the first one is the `atom-clock` package and the second one is `minimap`. Limited time to test right now because I am at work :slightly_frowning_face:. Just logged it here so I don't forget and in case anyone with more time wants to take a look before I have the chance. Maybe not even be related to this branch since I am comparing master to beta.

nathansobo commented:
I made some changes to the text editor in #13795 which could be causing the exception in question. I'd like to change the design of the editor before upgrading Electron because the scrollbars are behaving badly on the latest Chrome and I'd rather dispense with the complexity than double down on it.

50Wliu commented:
One more comment regarding subpixel AA.  Any line that is completely italicized will disable subpixel antialiasing on that line as well as surrounding lines (the exact number appears to be dependent on blank lines and other factors that I can't pinpoint at the moment, maybe tiling?).

Ben3eeE commented:
@nathansobo I added the minimap exception to the list in the body of the issue.

mybks commented:
@nathansobo I've compiled the tj-upgrade-electron, and the fonts are still not as sharp as in OS, why is so that? and maybe you know if it's going to change in the future in electron, or chromium? Atom has such amount of features, but everytime I switch to it I feel my eyes just start drying.

lee-dohm commented:
@mybks Chromium uses its own font rendering engine to provide cross-platform consistency, so the font rendering in Chromium, and by extension Atom, will not always match the OS font rendering on your platform. If the font rendering you see in Atom matches Chromium then that is the best we can do. You would have to take a look at the Chromium team's roadmap to see what their plans are for improving font rendering on your platform long-term.

scherii commented:
Just wanted to point out that there is a  `#enable-harfbuzz-rendertext` in [chrome://flags](chrome://flags) option that, as far as I understand, turns on the [HarfBuzz engine](https://www.wikiwand.com/en/HarfBuzz) (see https://github.com/behdad/harfbuzz).

thomasjo commented:
> Just wanted to point out that there is an option called #enable-harfbuzz-rendertext in chrome://flags that, as far as I understand, turns on the HarfBuzz engine (see https://github.com/behdad/harfbuzz).

No such option here using Chrome 56 (Ubuntu 16.04). From a cursory glance in the Chromium source, it looks like this is the default OT engine on (nearly) all supported platforms.

scherii commented:
Yeah, you're right, this is on Chrome Canary 59 on macOS.
The flag `#enable-harfbuzz-rendertext` is macOS-only.

mybks commented:
@lee-dohm Thank you very much for your response, will take a look on the roadmap. And thanks also @scherii for suggestion

nathansobo commented:
@thomasjo @50Wliu Based on some developments concerning v8 snapshots, I think it makes sense to go straight to Electron 1.6 as I have on the [`ns-electron-1.6.x` branch](https://github.com/atom/atom/tree/ns-electron-1.6.x).

Basically, snapshots are broken on all platforms for 1.4.6 and above and broken on Windows for all of 1.4.x. Snapshots work perfectly on 1.6 and the rewards of upgrading to 1.6.x are pretty high anyway. I'd like to save the Electron team the hassle of hotfixing 1.4 to support snapshots since it's an outdated branch.

You could reset this branch to that branch if you agree with me.

thomasjo commented:
> Based on some developments concerning v8 snapshots, I think it makes sense to go straight to Electron 1.6 as I have on the ns-electron-1.6.x branch.
> ...
> You could reset this branch to that branch if you agree with me.

Agreed — and done! :100:

Note that this is somewhat more risky in some sense, but worth it.

50Wliu commented:
Unfortunately snapshots aren't working for me on Windows after upgrading to Electron 1.6.  It seems to be the same error that appeared before upgrading to Electron 1.3.14.

kevinsawicki commented:
> Unfortunately snapshots aren't working for me on Windows after upgrading to Electron 1.6. It seems to be the same error that appeared before upgrading to Electron 1.3.14.

Yeah, we haven't cut an Electron 1.6.x release yet with this fix, should come out early next week.

nathansobo commented:
> Unfortunately snapshots aren't working for me on Windows after upgrading to Electron 1.6

Oops, sorry I misspoke. Maybe you can disable them if you're living on this branch until next week.

kevinsawicki commented:
> Yeah, we haven't cut an Electron 1.6.x release yet with this fix, should come out early next week.

This is on electron master now and will be included in the 1.6.4 beta next week.

Commit:
Render overlay decorations

Commit:
Support class property on overlay decorations

Commit:
Use spies instead of monkey patching

Commit:
Add avoidOverflow: false option for overlays

Commit:
Add static TextEditor.viewForOverlayItem method to avoid using global

Commit:
Use the atom.views scheduler in TextEditorComponent

This ensures smooth scheduling interactions with autocomplete-plus
overlays so they measure their dimensions at the right time.

Commit:
Refactor

Commit:
Factor editor component update into high-level phases

Commit:
Fix wrong variable name

Commit:
Integrate properly with Atom scheduler

Commit:
Only update scrollTop/Left when they change

This avoids forcing a reflows in some circumnstances.

Commit:
Start on custom gutters

Commit:
Associate gutters with their elements and support showing/hiding gutters

50Wliu commented:
Looks like we may need to update the CSP:
```
Refused to load the font 'atom://file-icons/fonts/fontawesome.woff2' because it violates the following Content Security Policy directive: "default-src *". Note that 'font-src' was not explicitly set, so 'default-src' is used as a fallback.
```

file-icons is completely broken as a result.
![broken-csp-file-icons](https://cloud.githubusercontent.com/assets/2766036/24229970/f1d8847a-0f52-11e7-8c98-d71856fda298.png)

50Wliu commented:
Shift-scrolling (to scroll horizontally) doesn't work.

Commit:
:arrow_up: etch to allow arbitrary objects as keys

Commit:
Adjust left position of scroll container when gutter container resizes

50Wliu commented:
Electron 1.6.x uses Node 7, which deprecated `os.tmpDir()`.  I investigated solving that deprecation warning, but ran into some trouble:
* Atom and its packages currently depend on either `temp` or `tmp` for creating temp files (`temp` is the package we use more)
* Both of those packages rely on `os-tmpdir`, which calls `os.tmpDir()`
* `os-tmpdir` is deprecated
* `temp`'s last release was 2 years ago
* `tmp` is striving for backwards-compatibility, all the way back to Node 0.6, and is reluctant to release a 1.0.0 version that will get rid of the `os-tmpdir` dependency

Commit:
WIP: Render gutters before initial measurement

The shouldUpdate method is just returning true for now. We probably need
to find a new approach to representing line number decorations that's
easier to diff, perhaps a sparse array?

ZahidDev commented:
@50Wliu Latest version of `node-tmp` at https://github.com/raszi/node-tmp works for me on node version 7.7.4. The os-tmpdir implementation used in the latest version, while depreciated, doesn't rely on `os.tempDir()` nor `os.tempdir()`. Thus being, backwards compatible. I don't seem to get depreciation warnings on Electron 1.6.x

50Wliu commented:
Ok, I'll take a look.  Thanks!

Commit:
Set the height of the line number gutter explicitly

Commit:
Store line, line number decorations in arrays and avoid slicing

Commit:
Cache subtrees to avoid duplicating work within a single frame

Commit:
Avoid using += with let variables to avoid let compound assigment deopt

See https://jsperf.com/let-compound-assignment

Commit:
Unfocus test

Commit:
:arrow_up: etch

Commit:
Blink cursors; still needs tests

Commit:
Implement a fast path for cursor blink to minimize battery impact

Commit:
Add smoke test for cursor blink

Commit:
Correctly assign gutter container vnode

Commit:
Ensure nextUpdateOnlyBlinksCursor argument is `true`, not just truthy

We pass the bound scheduleUpdate method as an event handler to a variety
of subscription methods, some of which supply arguments.

Commit:
Speed up cursor blink test

Commit:
Add setInputEnabled and don't handle textInput if it is disabled

Commit:
Test and fix the `center` option to autoscroll

Commit:
Cache prefixed scope names

as-cii commented:
The deprecation message should have been solved by e6c74d6. If you are still getting it, please delete your `node_modules` folder and rebuild: npm will resolve the correct dependencies and the deprecation notice will go away. This should happen automatically on CIs after this branch gets merged into master: the dependencies fingerprint will be updated as a result of the electron upgrade, and therefore the generated assets will contain the correctly resolved dependencies.

Ben3eeE commented:
I see this in the console when activating [minimap-autohider](https://atom.io/packages/minimap-autohider)

```
(node:21264) DeprecationWarning: Calling an asynchronous function without callback is deprecated.
process.on @ internal/process/warning.js:21
emitOne @ events.js:96
emit @ events.js:188
process.nextTick @ internal/process/warning.js:47
_combinedTickCallback @ internal/process/next_tick.js:67
_tickCallback @ internal/process/next_tick.js:98
```

as-cii commented:
> I see this in the console when activating minimap-autohider

After a quick inspection of `minimap-autohider`, it seems like they are not specifying a callback when calling `fs.writeFile` (https://github.com/ansballard/minimap-autohider/blob/master/lib/minimap-autohider.js#L64). Since it's just a deprecation I believe it's not necessary to provide a shim in Atom,  but it may still be good to open an issue on their repository and inform them of the deprecation.

ansballard commented:
`minimap-autohider` maintainer here. Not a problem, I'll try to get that patched in tonight. Are there docs on why a callback needs to be specified? I can throw a noop in real quick, but if there's some new standard I should be following I might as well do that.

Arcanemagus commented:
@ansballard You can find the documentation on `fs.writeFile` [here](https://nodejs.org/api/fs.html#fs_fs_writefile_file_data_options_callback). Note that the callback is marked as required. You would probably have to dig through the Node.js changelogs to find the reasoning for that though.

ansballard commented:
Yep, marked as required in `7.0.0`, guess I'll add in some error handling. Thanks for the info!

kevinsawicki commented:
You may want to disable to new `auxclick` event in this branch.

Chrome 55 shipped with it and it causes the `click` event to not fire anymore for middle clicks which I would guess would break some packages or core behavior that handles middle clicks explicitly. It also causes Electron to open the link in a new window when middle-clicked which is probably rarely the desired behavior.

You can disable it and go back to the pre-55 behavior by specifying the following option in the `BrowserWindow` constructor:

```js
webPreferences: {
  // Disable auxclick event
  // See https://developers.google.com/web/updates/2016/10/auxclick
  disableBlinkFeatures: 'Auxclick',
}
```

https://developers.google.com/web/updates/2016/10/auxclick
https://www.chromestatus.com/feature/5663174342737920

Commit:
WIP: Start on block decorations

bpasero commented:
@kevinsawicki thanks for sharing, we just hit this issue too. For `webviews` the relevant thing to do seems to be: `webview.setAttribute('webPreferences', 'disableBlinkFeatures=Auxclick');`

kevinsawicki commented:
> What I am not sure is: If you set this in webview, are the defaults being overridden? And if so, what are the defaults?

The defaults would still apply and whatever you specify is only additive (similar to the `BrowserWindow` constructor).

You could also use: https://electron.atom.io/docs/api/webview-tag/#disableblinkfeatures

```html
<webview src="https://www.github.com/" disableblinkfeatures="Auxclick"></webview>
```

Commit:
Use `LineTopIndex` to convert from/to rows to/from pixel positions

Commit:
Add fast path when no block decorations need to be measured

Commit:
Use `Array.push` instead of `array[i] =` when adding line components

Albeit (potentially) slower, this will allow to add a dynamic number of
block decoration nodes before and after a given line.

Commit:
Render block decorations between lines

Commit:
Test scrolling down with block decorations

Commit:
Handle inserting and updating block decorations

Commit:
Splice `LineTopIndex` when a textual change occurs

Commit:
Don't use etch for lines rendering

50Wliu commented:
I'm seeing the dev tools freeze again after reloading the window.  Steps to reproduce seem to be:
1. `atom --dev`
2. Open dev tools
3. Reload window
4. Notice that the dev tools, while not frozen, doesn't update and is essentially useless

Commit:
Integrate block decorations in the custom lines rendering routine

Commit:
Align line number nodes with line nodes

Commit:
Add TextEditorElement.prototype.invalidateBlockDecorationDimensions

Commit:
Hide cursors with non-empty selection if showCursorsOnSelection is false

Also, remove some barely used public APIs around cursor visibility that
don't make much sense and are not ideal for performance. We don't want
to subscribe to the visibility of each cursor.

Commit:
Update line number gutter when invalidating a visible block decoration

When two or more decorations in the same tile were invalidated but the
sum of their height didn't change, we were previously failing to
recognize that the line numbers gutter needed to be re-rendered. With
this commit, whenever a block decoration is visible and gets
invalidated, we will force the line number gutter to always update.

Commit:
Remeasure block decorations when editor width changes

Signed-off-by: Nathan Sobo <nathan@github.com>

Commit:
Fix typo

Commit:
Re-measure and update rendered content when editor styles change

Commit:
Don't update editor component if we know we are not visible

Commit:
Remove stray logging

Commit:
Don't re-measure if editor has become invisible

Commit:
Fix test

Commit:
Clear highlight nodes when recycling line tiles

Commit:
Include more properties in LinesTileComponent.shouldUpdate

Commit:
Remove unused var

Commit:
Remove highlight caching for now

Commit:
Slice lines and decorations passed to LinesTileComponent

This ensures the component's shouldUpdate method works correctly.

Commit:
Don't count the dummy line in block decoration test

Commit:
Fix bad syntax in src/initialize-benchmark-window.js

Commit:
Import octicon-mixins in static/text-editor.less

Commit:
Fix lint errors and delete dead code

Commit:
Further optimize line replacement

Commit:
Wait until the editor is focused before starting to type in smoke test

Commit:
Update fake gutter container interface in gutter-spec.coffee

Commit:
Stop calling `initialize` in `ViewRegistry` tests

Commit:
Don't reuse resize detectors across `TextEditorComponent` instances

Due to the way element-resize-detector schedules the delivering of
resize events, this will ensure that creating an editor while the clock
is mocked won't prevent subsequent tests using the real clock from
getting such events.

Commit:
Test clicking more locations outside of the lines

Commit:
Fix spec/text-editor-spec.coffee

Commit:
Register style elements change events in AtomEnvironment.initialize

...and fix spec/workspace-element-spec.js

Commit:
Create resize detector before calling etch.updateSync for the first time

Commit:
Skip obsolete tests for now, but delete them later

Commit:
Add data-grammar to editor element

Commit:
Add data-encoding to editor element

Signed-off-by: Nathan Sobo <nathan@github.com>

Commit:
Add deprecated shadow root property to editor element

Signed-off-by: Nathan Sobo <nathan@github.com>

Commit:
Don't share block decoration/character measurement vnodes across instances

Signed-off-by: Nathan Sobo <nathan@github.com>

Commit:
Position dummy line element absolutely and make it invisible

Signed-off-by: Nathan Sobo <nathan@github.com>

Commit:
Replace element-resize-detector with experimental ResizeObserver API

Signed-off-by: Nathan Sobo <nathan@github.com>

Commit:
Don't use position: relative on atom-text-editor

Commit:
Account for scrollbars and padding in autoHeight/Width mode

Commit:
Hide scrollbars in mini editors

Commit:
Fix clicking fold placeholders by ignoring pointer events on cursors div

Commit:
Give cursors at the end of lines the width of an 'x' character

Commit:
Add has-selection class to editors with non-empty selections

Commit:
Delete obsolete code and tests related to text editor rendering

Commit:
Get TextEditorElement tests passing

Commit:
Fix lint errors

Commit:
Set lineHeightInPixels on model for backward compatibility

Commit:
Replace getDefaultCharacterWidth with getBaseCharacterWidth

That's the language we use throughout the implementation now and a more
accurate name for the concept.

Commit:
Convert text-editor-element-spec to JS

Commit:
Use async/await in text-editor-element-spec

Commit:
Add TextEditorElement.pixelPositionForScreen/BufferPosition

These methods require us to render off-screen lines in some
circumstances in order to measure them, so this commit extends the
rendering of the longest line to include arbitrary lines.

Commit:
Add deprecated rootElement property

Commit:
Round column measurements to nearest whole pixel

This preserves the expected behavior for positioning overlays, etc so
that package tests keep passing.

Commit:
Assign bufferRow property to line number nodes

I wish we didn't need this, but it's currently relied on by several
packages including bookmarks.

Commit:
Don't pane focus when pane model is destroyed

This avoids a non-failure error message when resetting the environment
in some specs.

Commit:
Honor the `updateSynchronously` parameter

Commit:
Disconnect resize observers in overlay components on editor detach

Signed-off-by: Nathan Sobo <nathan@github.com>

Commit:
Assign placeholder text on the model only when the attribute is present

Signed-off-by: Nathan Sobo <nathan@github.com>

Commit:
Disable ResizeObserver temporarily in resize callback to avoid warning

Signed-off-by: Nathan Sobo <nathan@github.com>

Commit:
Add keys to gutterContainer and scrollContainer to avoid recycling issue

Previously, when the gutter container was removed due to the editor
becoming mini, the lack of keys caused the gutter to be updated with the
recycled cursors vnode. But then we tried to remove the cursors vnode
not realizing it had been moved and tore down all the references.

We probably need to revisit whether it makes sense to recycle vnodes.

Commit:
Delegate (get|set)(Height|Width) to element

Rather than storing these values on the editor model.

Commit:
Add back the measureDimensions method since some packages rely on it

Ideally, packages would resize and then wait for an update. But we set
up an example of calling measureDimensions directly in find-and-replace
so the easiest thing for now is just to keep this method around.

Commit:
Delegate getFirst/LastVisibleScreenRow from model to component

Commit:
Delegate setFirstVisibleScreenRow from the model to the component

Commit:
Delegate get/setFirstVisibleScreenColumn from the model to the component

Commit:
Guard gitFirst/LastVisibleScreenRow

These methods are sometimes called by the model before the editor has
been attached to the DOM.

Commit:
Drop tests for set/getFirstVisibleScreenRow

These are now tested in text-editor-component-spec

Commit:
Return false from isLineCommentedAtBufferRow if no line yet exists

Commit:
Set updatedSynchronously to false in text-editor-element-spec

Commit:
Don't update synchronously in text-editor-element-spec

Commit:
Add tests for TextEditorElement.setScrollTop/Left

Commit:
Restore editor scroll position across reloads

This commit introduces the concept of a scrollTopRow and
scrollLeftColumn which is used to query and update the logical scroll
position.

Commit:
Use `position: relative` for `.line-number` elements

...because packages like `.git-diff` are relying on this behavior to
position their decorations. This didn't seem to degrade layout times, so
it makes sense to just add it to keep package breakage to a minimum.

Commit:
Honor the gutter-hidden attribute correctly

Commit:
Add better test coverage for the mini and placeholder-text attributes

Commit:
Remeasure gutter dimensions when a gutter changes its visibility

Commit:
Implement `TextEditor.prototype.getRowsPerPage`

Commit:
Don't remove non accented character from history, improve test coverage

Unfortunately Chromium does not trigger a `compositionstart` before
firing the text input event for the non accented character. Using `undo`
to remove such character from the history is risky because it could be
grouped with a previous change, thus making Atom undo too much.

With this commit we simply keep the behavior master exhibits as of
today. In the process of rewriting this code path, however, we fixed a
bug that occurred when opening the accented character menu while holding
another key, and improved test coverage as well by simulating the events
the browser triggers.

bronson commented:
Another issue fixed by this PR: https://github.com/electron/electron/issues/7403

Right now, both Atom channels (1.16 and 1.17-beta) produce 0-length files when saving a CPU profile.

50Wliu commented:
Thanks, I've added the Atom issue to the fixed list.

Commit:
Fix pageUp/Down tests by using a real element

Commit:
Only enable cursor blink optimization when updateSync is using scheduler

This ensures that direct calls to updateSync from places like scroll
handlers never take this optimization path.

Commit:
Fix lint error

Commit:
Don't render decorations for invalidated markers

Commit:
Make process.platform easier to mock

Commit:
Add middle mouse pasting on Linux

Commit:
Fix cursor positioning around fold markers

Commit:
Add data-screen-row to line nodes

Commit:
Schedule update when setting scroll top row or scroll left column

Commit:
Add cursor decorations

These decorations allow the class and style of a cursor associated with
any marker to be customized.

/cc @t9md

Commit:
Add TextEditorComponent.screenPositionForPixelPositionSync

This method can be used to translate a pixel position to a screen
position even if the line is not currently rendered on screen.

Commit:
Give line numbers the full width of the line number gutter

Commit:
Consolidate editor style sheets

Commit:
Drop floats

Commit:
Use `Math.round` for positions that are at the end of a line

Commit:
Schedule component updates directly from the model

The only event-based APIs we kept are for listening to changes in block
decoration markers.

Commit:
Fix more lint errors

Commit:
Update mock text editor in gutter-container-spec.coffee

Commit:
Fix shift-scroll on Windows and Linux

Commit:
Restore scrollbar positions correctly on reload

Commit:
Add {get,set}FirstVisibleScreen{Row,Column} to TextEditorElement

Commit:
Provide an `editorElement` shim on TextEditor

Signed-off-by: Nathan Sobo <nathan@github.com>

Commit:
Add a `pixelPositionForMouseEvent` method

This was a private method in the previous implementation that was used
by some packages.

Signed-off-by: Nathan Sobo <nathan@github.com>

Commit:
Use `cursor:text` on atom-text-editor elements

Signed-off-by: Nathan Sobo <nathan@github.com>

Commit:
Don't remeasure scrollbars for mini editors

Signed-off-by: Nathan Sobo <nathan@github.com>

Commit:
Add highlight decoration classes to region elements as well

Signed-off-by: Nathan Sobo <nathan@github.com>

Commit:
Add highlights class for package compatibility

Commit:
Add scrollbar classes in case any packages or themes target them

Commit:
Remove some redundant styling

Commit:
Add missing methods on TextEditorElement

Commit:
Add intersectsVisibleRowRange on TextEditorElement

Commit:
Add TextEditorElement.pixelRectRangeForScreenRange for compatibility

Commit:
WIP: Start extracting gutter component

Commit:
Initialize all measurements to 0

Commit:
:art:

Commit:
Pass props to GutterContainerComponent instead of reaching up to parent

There are still a few rootComponent references remaining in the
LineNumberGutterComponent. These should be removed and instead we should
consult this data when constructing the line numbers to render.

Commit:
Eliminate caching of linesVnode

Commit:
Extract CursorsAndInputComponent

Commit:
Eliminate cached placeholderTextVnode

Commit:
Update gutter container only once per frame unless its width changes

Signed-off-by: Nathan Sobo <nathan@github.com>

Commit:
Update cursors only once per frame (after content has been measured)

Signed-off-by: Nathan Sobo <nathan@github.com>

Commit:
Update lines content only once per frame

Signed-off-by: Nathan Sobo <nathan@github.com>

Commit:
Cache derived dimensions during each phase of `updateSync`

Signed-off-by: Nathan Sobo <nathan@github.com>

Commit:
Fix build failures

Commit:
Fix positioning for block decorations located at the beginning of a tile

Commit:
Set autoHeight: true explicitly in benchmarks

Commit:
Eliminate strictly contained divs wrapping lines and highlights

I was hoping to strictly contain the layouts of highlights an lines
separately, since they are updated during different render phases.
Unfortunately, strict containment requires both divs to be positioned
absolutely. This in turn creates separate stacking contexts for lines
and highlights, which makes it impossible to render highlights in front
lines which themes sometimes need to do. For example,
atom-material-syntax pushes bracket matcher highlights to the front so
they are not obscured by the theme's solid black cursor line background.

/cc @as-cii. You should examine my work here and make sure I'm not
screwing something up with your line/block decoration update code.

Commit:
Fix error when attaching soft-wrap editor in synchronous update mode

Taking the initial measurement was setting the soft wrap column, which
was triggering a display layer reset, which was scheduling an update.
This update occurred at an unexpected time causing an exception.

Commit:
Avoid blowing away classes assigned on the editor element by packages

/cc @t9md

Commit:
Fix lint errors

Commit:
Ignore resize events if they are delivered while the editor is hidden

Commit:
Use scope ids instead of tags

Commit:
Implement text decorations in rendering layer

Commit:
Never create empty spans at the beginning of a row

This was happening when a text decoration overlapped a row, but the next
boundary was located exactly at the beginning of it.

Commit:
Don't activate scrollPastEnd for autoHeight editors

Commit:
Make first-mate scope ids always larger than built-in ones

Commit:
:arrow_up: text-buffer

Commit:
Update DOM in screenPositionForPixelPosition if needed

Some packages are interacting with this method assuming this behavior,
so this commit eliminates `screenPositionForPixelPositionSync` and
instead just performs the DOM update in `screenPositionForPixelPosition`
if it is needed.

Commit:
Account for vertical scrollbar width when soft-wrapping lines

Commit:
Set `overflow: hidden` and `contain: layout paint` on lines

Commit:
Pass mini attribute when creating new TextEditor from TextEditorElement

This avoids content being shifted over due to rendering and measuring
the gutter on element creation and then subsequently hiding it.

Commit:
Rename method to match old implementation

Commit:
Avoid requesting horizontal measurement when auto-scrolling vertically

This was leaving a measurement request in the map that was getting
picked up on the next frame. In some cases, the requested measurement
row was not present, causing an exception.

Commit:
Only resolve update promise after final render phase

Commit:
Maintain the scroll position when changing font size

Commit:
Fix remaining test failures in core

Commit:
Fix remaining linting warnings

Commit:
Update class list even when the editor is not attached

Commit:
Don't create empty nodes when a text decoration ends next to a text tag

This was causing problems in measurements because in that code path we
assume that text nodes are never empty. This commit also adds a test
verifying this invariant when a text decoration ending right after a
text tag is added.

as-cii commented:
@50Wliu: I can't reproduce the issue you reported regarding dev tools freezing after reloading in dev mode (I tested it on #13880). I marked it as completed here, but feel free to uncheck it if you can still reproduce.

Commit:
Don't 'contain: paint' line elements

This fixes
https://github.com/atom/atom/pull/13880#issuecomment-296623782 once
again.

Commit:
Fix clicking past the content height

Signed-off-by: Nathan Sobo <nathan@github.com>

Commit:
Update width of content when approximate longest screen row changes

Stanzilla commented:
don't forget to bump this to 1.6.8, it has some good bug fixes

Commit:
Remove dead code

Commit:
Move cursors within a transaction to batch marker layer update events

Signed-off-by: Nathan Sobo <nathan@github.com>

Commit:
Call `editor.setEditorWidthInChars` only when the value changed

Signed-off-by: Nathan Sobo <nathan@github.com>

Commit:
Merge pull request #13880 from atom/ns-editor-rendering

Rewrite editor rendering layout to use new browser features and virtual DOM

Commit:
Delete more dead code

Commit:
Fix rendering artifacts when resizing with soft wraps

Previously, we were accidentally depending on the state of the display
layer when forcing it to update its index. This caused us to not index
enough content to cover the visibile area, which meant we weren't
querying enough lines to fill the screen in some situations.

Commit:
:arrow_up: text-buffer

maxbrunsfeld committed:
:racehorse: Ensure rendered tile count is stable when scrolling

Signed-off-by: Nathan Sobo <nathan@github.com>

Commit:
Don't populate more screen rows than necessary

See 1c8847cb4f4ce46f3639471197e8e552999fd19b

Commit:
Use explicit state to recycle tiles instead of modulo scheme

This avoids updating tiles unnecessarily when changing the number of
rendered tiles.

Signed-off-by: Nathan Sobo <nathan@github.com>

Commit:
Null out component when editor is destroyed to minimize leaks

Signed-off-by: Nathan Sobo <nathan@github.com>

Commit:
Dispose nodes associated with a LineComponent when replacing it

Commit:
:arrow_up: etch

Commit:
Don't change number of tiles based on block decorations

This means we may render more tiles than necessary when we have block
decorations, but it prevents changing the number of rendered tiles
during scrolling with certain combinations of line height and editor
height. If it ever becomes a problem we can get smarter about
subtracting the height of the visible block decorations from the editor
height, but for now this gives us more reliable performance for the
common case.

Commit:
Avoid deopt when calling TokenizedLine.prototype.getTokenIterator

Commit:
Don't populate more screen rows than necessary

See 4eecf8d1a683e6b4e4c3b16a08907b1bfd7f0af7.

Commit:
Use bufferRowsForScreenRows to compute line numbers

Commit:
:arrow_up: line-top-index

Commit:
:arrow_up: etch

Commit:
Ignore mismatched tags between lines in TokenizedBufferIterator

Previously the DisplayLayer would have issues when closing tags that had
not been previously opened, requiring us to guard against mismatched
scopes at the frontier of asynchronous tokenization (see 5cfe9716 for
more information).

Now the DisplayLayer gracefully handles closing tags that had not been
opened, meaning we can eliminate this costly logic.

Signed-off-by: Nathan Sobo <nathan@github.com>

Commit:
:arrow_up: text-buffer

Commit:
Optimize isFoldable

Signed-off-by: Nathan Sobo <nathan@github.com>

Commit:
Avoid creating a new timeout every time cursor blinking is paused

Commit:
Fix blurring the editor when no hidden input is present

Commit:
Change editor.bufferRowsForScreenRows to include the supplied endScreenRow

Commit:
Re-enable a test that was previously failing

Commit:
Delete unused code

Commit:
:lipstick:

Commit:
Merge branch 'master' into tj-upgrade-electron

# Conflicts:
#	script/package.json
#	src/task.coffee

Commit:
:arrow_up: electron

Commit:
:arrow_up: text-buffer

Commit:
:lipstick: Fix more lint errors

Commit:
Use buffer coordinates in isFoldedAtCursorRow

This avoid an extra screen-to-buffer coordinate translation.

Signed-off-by: Nathan Sobo <nathan@github.com>

Commit:
Don't clip screen range when autoscrolling from Cursor

Signed-off-by: Nathan Sobo <nathan@github.com>

Commit:
Prevent unnecessary clipping of buffer range in isFoldedAtBufferRow

Signed-off-by: Nathan Sobo <nathan@github.com>

Commit:
:arrow_up: first-mate

Commit:
Mock Date.now in spec helper in addition to _.now

Signed-off-by: Max Brunsfeld <maxbrunsfeld@gmail.com>

Commit:
Disable github package in smoke test – it takes too long to compile

Commit:
Eliminate reliance on local storage in HistoryManager

It's causing test failures locally and enough time has passed that most
user data should be transitioned to indexed DB by now.

Commit:
Avoid problems related to calling `startEditorWindow` in tests

Commit:
Avoid polluting test with click events on the test runner

We now pass an isolated document into the AtomEnvironment instance under
test to avoid accidentally handling clicks on the test runner itself.

Commit:
Destroy environment to silence leaked editor warning

Commit:
Fix atom.setSize spec when window is full screen

Commit:
Silence error output in test

Commit:
Fix lint errors

Commit:
Unspy Date.now when jasmine.useRealClock is called

Commit:
Upgrade packages so tests pass with a mocked Date.now

Commit:
Fix editor component tests on Windows

Commit:
:arrow_up: first-mate

Commit:
Fix editor element tests on Windows

50Wliu commented:
Seeing some odd folding behavior.
When soft-wrapped:
1. Folding indicator does not update state.
2. Incorrect folding indicator at the end of the soft-wrapped line that doesn't do anything.
![folding-incorrect-state](https://cloud.githubusercontent.com/assets/2766036/26012999/7b56b4da-3724-11e7-9630-de4755ae6fc2.gif)

Unrelated to soft-wrap:
3. Unable to unfold selection
![folding-unable-to-unfold](https://cloud.githubusercontent.com/assets/2766036/26013105/eb03bc60-3724-11e7-8bef-db5f2b48c515.gif)

Arcanemagus commented:
> Failed to activate the pigments package, cannot find highlights-component in the require cache We'll need to introduce a new API to support the use case that @abe33 is monkey patching private implementation details to support. We'll need to merge this before doing that.

That was marked as checked in the rendering re-write, but I can't find a related issue in the `pigments` repo, and didn't see a related issue filed here. Is that being tracked somewhere?

Arcanemagus commented:
If the toolbar is hidden and temporarily shown, when it hides again there is a brief white flash as the interface is moved up to take over the newly free space but hasn't yet redone the layout:
![2017-05-12_13-44-57](https://cloud.githubusercontent.com/assets/427137/26016320/6225f42c-3719-11e7-80c4-99ff9173219e.gif)
Reproduces virtually every time, the GIF only captures a few instances though. Note that Atom v1.17.0-beta5 has the same delay in rendering, but the blank area that shows briefly is the same color as the rest of the app background leading to it not being as noticeable.

nathansobo commented:
@Arcanemagus I checked the box because I added the raw API to *support* pigments, but perhaps I shouldn't have. Could you maybe create an issue on the pigments repository and cross-link it? I'm pretty sure @as-cii had a conversation with @abe33 about this and he's on board to switch to the new APIs, but may not have had time yet.

As for the other bug, noted. Ill add it to the check list. Since it's not a deal-breaker we may merge this before we fix it.

Arcanemagus commented:
I think I'm seeing a related issue to what @50Wliu reported in https://github.com/atom/atom/pull/12696#issuecomment-292690910.

1. Open Atom in safe mode and open dev tools
2. Reload Atom (<kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>F5</kbd>)
3. Get a blank window

Clicking in the blank window shows this error in the console of the open Dev tools:
```Uncaught Error: Pane has been destroyed
    at Pane.module.exports.Pane.activate (C:\Temp\atom\out\Atom x64\resources\app\src\pane.js:872:1)
    at HTMLElement.handleFocus (C:\Temp\atom\out\Atom x64\resources\app\src\workspace-element.js:222:32)
module.exports.Pane.activate @ <embedded>:29041
handleFocus @ C:\Temp\atom\out\Atom x64\resources\app\src\workspace-element.js:222
```
![image](https://cloud.githubusercontent.com/assets/427137/26020161/33448ce2-3730-11e7-8c39-5c7c65a74e10.png)

Even stranger, attempting to close Dev tools causes it to immediately re-appear with another of those errors in the Console.

Also very likely related, but slightly different: If you try to take a CPU profile and while recording reload the Atom window you get an "Editor has crashed" dialog (and the same blank window).

Arcanemagus commented:
Also on first opening a window (with a saved session state) I see this in the console:
![image](https://cloud.githubusercontent.com/assets/427137/26020291/89b60e9c-3731-11e7-959d-cf6b313ea3e6.png)
```
IntersectionObserver.observe(target): target element is not a descendant of root.
TreeView @ <embedded>:105039
```

Ben3eeE commented:
@Arcanemagus Thanks, adding this to the list in the body of the PR. Do you see the dev tools reload misbehaving in safe mode?

Arcanemagus commented:
> Open Atom in **safe mode** and open dev tools

@Ben3eeE Yep 😛.

Arcanemagus commented:
Can confirm both of the above on a build from https://github.com/atom/atom/pull/12696/commits/96533998e2da5ee8f3adadc86efb48e9405ae6e5:
```
OS: Windows 10 x64
Atom    : 1.18.0-dev-96533998e
Electron: 1.6.9
Chrome  : 56.0.2924.87
Node    : 7.4.0
```

50Wliu commented:
Not too sure what happened or if this is isolated to the Electron upgrade but my Find menu looks like this now:
![broken-find-menu](https://cloud.githubusercontent.com/assets/2766036/26078215/c3c59262-398c-11e7-878f-4519c3cb9511.png)

Tentative repro steps seem to be opening the menu through the <kbd>alt-i</kbd> accelerator.

steelbrain commented:
Just saw a little lag on this branch, can anyone else repo? (filesize was too large so had to upload to Google Drive)

https://drive.google.com/file/d/0B_aQKK-B3pCfQjBncWFGcndfQTA/view

Commit:
Update foldable icon when a row's foldability changes

Commit:
Show foldable icon on the last screen row belonging to a buffer row

Commit:
Always allow to destroy free-form folds from the gutter

Commit:
Merge branch 'master' into tj-upgrade-electron

Commit:
:arrow_up: text-buffer

as-cii commented:
> Just saw a little lag on this branch, can anyone else repo? (filesize was too large so had to upload to Google Drive)

I couldn't reproduce the find-and-replace issue, but it might be worth to keep an eye on it and make sure it didn't regress with these changes. I am pretty sure it's not related to the editor rendering rewrite though.

> Not too sure what happened or if this is isolated to the Electron upgrade but my Find menu looks like this now:
> 
> Tentative repro steps seem to be opening the menu through the alt-i accelerator.

Interesting, unfortunately this is not happening for me on macOS. Can you reproduce consistently? @Ben3eeE: if I recall correctly you use Windows as well. Can you reproduce this on your machine?

> Also very likely related, but slightly different: If you try to take a CPU profile and while recording reload the Atom window you get an "Editor has crashed" dialog (and the same blank window).

I have noticed this one too lately, it might have something to do with the new Electron. Interestingly enough, reloading when DevTools are closed works perfectly. I'll file a bug report on Electron and see if we can fix this somehow. 

> Seeing some odd folding behavior.

This should be fixed as of the latest commit. Nice finding! ✨

Commit:
:art:

Commit:
:arrow_up: bracket-matcher

Arcanemagus commented:
> Not too sure what happened or if this is isolated to the Electron upgrade but my Find menu looks like this now:

Just a note I couldn't reproduce this either (on Windows).

Ben3eeE commented:
> @Ben3eeE: if I recall correctly you use Windows as well. Can you reproduce this on your machine?


I am unable to reproduce on Windows 7.

kaiser101 commented:
How long before this change will be merged into the master?

50Wliu commented:
I can't reproduce the find menu issue in safe mode, so it looks like that one was a false alarm.

nathansobo commented:
@kaiser101 We're facing some hard crashes with Electron when the dev tools are open. We may merge anyway and figure it out on master though.

nathansobo commented:
@steelbrain Unclear from your gif... Were you seeing lag scrolling the editor? If so, the lag could be due to longer pauses garbage collecting DOM nodes as we dropped the recycling strategy. @as-cii I wonder if we should revisit that. At some point we're gonna GC DOM and I worry those pauses could get long. It allocates quite a bit of memory. But first priority is these crashes I suppose.

steelbrain commented:
@nathansobo The editor was scrolling perfectly, the lag was only visible when scrolling search-and-replace tab (there would be white spaces in screen that would be filled later)

nathansobo commented:
@steelbrain That's definitely a find and replace bug and is worth reporting on that package if it hasn't been already.

Commit:
Use custom snapshot row translation to prevent Electron 1.6 reload crash

Adding a source map for the entire snapshot was expensive in terms of
memory and seemed to be triggering some sort of bug in Chromium when
reloading with the DevTools open.

The custom row translation relies on a much more compact representation
of the data and avoids the crash.

Signed-off-by: Nathan Sobo <nathan@github.com>

Commit:
Merge branch 'master' into tj-upgrade-electron

Commit:
Fix lint errors

Signed-off-by: Nathan Sobo <nathan@github.com>

as-cii commented:
Seems like AppVeyor failed but I believe that's just spurious. I will go ahead and merge it into master so that it can bake for a while before putting it on beta.

Thanks everyone for the help and the hard work on this! 🙏
